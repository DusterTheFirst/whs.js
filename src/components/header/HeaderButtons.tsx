/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import React from "react";
import { GestureResponderEvent, StyleSheet, Text, TouchableOpacity } from "react-native";

const styles = StyleSheet.create({
    button: {
        color: "#2f95dc",
        fontSize: 17
    },
    buttonContainer: {
        marginHorizontal: 15,
        marginVertical: 10,
        position: "absolute",
    },
    disabled: {
        color: "#dadada"
    },
    doneButton: {
        fontWeight: "bold"
    }
});

interface IHeaderDoneButtonProps {
    disabled?: boolean;
    onPress(event: GestureResponderEvent): void;
}
export function HeaderDoneButton({ onPress, disabled }: IHeaderDoneButtonProps) {
    return (
        <TouchableOpacity style={[styles.buttonContainer, { right: 0 }]} onPress={onPress} disabled={disabled}>
            <Text style={[styles.button, styles.doneButton, disabled === true ? styles.disabled : undefined]}>Save</Text>
        </TouchableOpacity>
    );
}

interface IHeaderBackButtonProps {
    onPress(event: GestureResponderEvent): void;
}
export function HeaderBackButton({ onPress }: IHeaderBackButtonProps) {
    return (
        <TouchableOpacity style={[styles.buttonContainer, { left: 0 }]} onPress={onPress}>
            <Text style={styles.button}>Cancel</Text>
        </TouchableOpacity>
    );
}