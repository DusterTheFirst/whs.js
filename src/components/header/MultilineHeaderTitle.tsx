/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import React, { memo } from "react";
import { GestureResponderEvent, TouchableOpacity } from "react-native";
import { HeaderSubtitleText, HeaderTitleText } from "../../styles/components/header";

/** The props for a multiline header */
interface IMultilineHeaderProps {
    /** The header title */
    title: string;
    /** The header subtitle */
    subtitle: string;
    /** The header click callback */
    onClick?(event: GestureResponderEvent): void;
}

/** A header title with multiple lines. A Title and a subtitle */
function MultilineHeaderTitle({ title, subtitle, onClick }: IMultilineHeaderProps) {
    return (
        <TouchableOpacity onPress={onClick}>
            <HeaderTitleText>{title}</HeaderTitleText>
            <HeaderSubtitleText>{subtitle}</HeaderSubtitleText>
        </TouchableOpacity>
    );
}

export default memo(MultilineHeaderTitle);