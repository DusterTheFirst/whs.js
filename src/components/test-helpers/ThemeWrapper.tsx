/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import React from "react";
import { ThemeProvider } from "styled-components/native";
import { lightTheme } from "../../styles/theme";

/** The wrapper around a tested component to provide it with a theme */
const ThemeWrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={lightTheme}>{children}</ThemeProvider>
);

export default ThemeWrapper;