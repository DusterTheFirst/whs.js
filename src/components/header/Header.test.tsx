/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import React from "react";
import { fireEvent, render } from "react-native-testing-library";
import ThemeWrapper from "../test-helpers/ThemeWrapper";
import { HeaderCancelButton, HeaderLeftArrow, HeaderRightArrow, HeaderSaveButton } from "./HeaderButtons";
import MultilineHeaderTitle from "./MultilineHeaderTitle";

describe("Tests different header buttons", () => {
    it("Renders <HeaderLeftArrow/>", () => {
        const mock = jest.fn();

        const comp = render(
            <HeaderLeftArrow onPress={mock} />,
            { wrapper: ThemeWrapper }
        );

        expect(comp.toJSON()).toMatchSnapshot();

        fireEvent.press(comp.getByA11yLabel("LeftArrow"));
        expect(mock).toHaveBeenCalledTimes(1);
    });

    it("Renders <HeaderRightArrow/>", () => {
        const mock = jest.fn();

        const comp = render(
            <HeaderRightArrow onPress={mock} />,
            { wrapper: ThemeWrapper }
        );

        expect(comp.toJSON()).toMatchSnapshot();

        fireEvent.press(comp.getByA11yLabel("RightArrow"));
        expect(mock).toHaveBeenCalledTimes(1);
    });

    it("Renders <HeaderCancelButton/>", () => {
        const mock = jest.fn();

        const comp = render(
            <HeaderCancelButton onPress={mock} />,
            { wrapper: ThemeWrapper }
        );

        expect(comp.toJSON()).toMatchSnapshot();

        fireEvent.press(comp.getByText("Cancel"));
        expect(mock).toHaveBeenCalledTimes(1);
    });

    it("Renders <HeaderSaveButton/>", () => {
        const mock = jest.fn();

        const comp = render(
            <HeaderSaveButton onPress={mock} />,
            { wrapper: ThemeWrapper }
        );

        expect(comp.toJSON()).toMatchSnapshot();

        fireEvent.press(comp.getByText("Save"));
        expect(mock).toHaveBeenCalledTimes(1);
    });
});

describe("Tests different variations of <MultilineHeader>", () => {
    it("Renders <MultilineHeaderTitle title=\"test title\" subtitle=\"e\"/>", () => {
        const comp = render(
            <MultilineHeaderTitle title="test title" subtitle="e" />,
            { wrapper: ThemeWrapper }
        );

        expect(comp.toJSON()).toMatchSnapshot();
    });
});