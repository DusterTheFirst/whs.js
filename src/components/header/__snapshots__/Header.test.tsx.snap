// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests different header buttons Renders <HeaderCancelButton/> 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#1f85cc",
          "fontSize": 17,
        },
      ]
    }
  >
    Cancel
  </Text>
</View>
`;

exports[`Tests different header buttons Renders <HeaderLeftArrow/> 1`] = `
<View
  accessibilityLabel="LeftArrow"
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
      "height": 30,
      "justifyContent": "center",
      "opacity": 1,
      "width": 30,
    }
  }
>
  <Text />
</View>
`;

exports[`Tests different header buttons Renders <HeaderRightArrow/> 1`] = `
<View
  accessibilityLabel="RightArrow"
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
      "height": 30,
      "justifyContent": "center",
      "opacity": 1,
      "width": 30,
    }
  }
>
  <Text />
</View>
`;

exports[`Tests different header buttons Renders <HeaderSaveButton/> 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#1f85cc",
          "fontSize": 17,
          "fontWeight": "bold",
        },
      ]
    }
  >
    Save
  </Text>
</View>
`;

exports[`Tests different variations of <MultilineHeader> Renders <MultilineHeaderTitle title="test title" subtitle="e"/> 1`] = `
<View
  accessible={true}
  focusable={false}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontSize": 15,
          "fontWeight": "bold",
          "textAlign": "center",
        },
      ]
    }
  >
    test title
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#808080",
          "fontSize": 12,
          "textAlign": "center",
        },
      ]
    }
  >
    e
  </Text>
</View>
`;
